src/python/WMCore/Services/DDM/DDM.py

```
@@ -0,0 +1,309 @@
+            So far it is working basically for the 'copy' API and due to that
+            the aggregation key is always a dictionary so no explicit check is
+            done for the type. In case we decide to use it we should develop
+            this to the end with the proper type checks.
+
+        :ddmReqList: List of objects, instances of DDMReqTemplate class.
+        :aggKey:    String representing the Aggregation key.
+        :return:     List of Dictionaries containing the results from DDM
+
+        """
+        if not ddmReqList or not isinstance(ddmReqList, list):
+            return None
+
+        ddmReqListAgg = []
+
+        # populate the first element in the aggregated list
+        ddmReqListAgg.append(ddmReqList.pop())
+
+        # feed the rest
+        while len(ddmReqList) != 0:
+            ddmReq = ddmReqList.pop()
+            found = False
+            for aggReq in ddmReqListAgg:
+                if ddmReq.isEqual(aggReq, aggKey):
+                    # Check if the two objects are not references to one and the
+                    # same object. Only then copy the values of the aggKey,
+                    # otherwise we will enter an endless cycle.
+                    if ddmReq is not aggReq and aggKey is not None:
+                        for i in ddmReq[aggKey]:
+                            aggReq[aggKey].append(i)
+                    found = True
+                    del(ddmReq)
+                    break
+            if not found:
+                ddmReqListAgg.append(ddmReq)
+
+        if not self.enableDataPlacement:
+            return ddmReqListAgg
+
+        results = []
+        for ddmReq in ddmReqListAgg:
+            api = ddmReq.api
+            if not api:
+                msg = "WARNING: skip sending data to DDM: "
+                msg += "API: {}: ddmReq: {}\n".format(ddmReq.api, pformat(ddmReq))
+                self['logger'].debug(msg)
+                results.append(None)
+                continue
```

This line here splits the actions from `makeRequest' in two steps, because in case of an error with some of the requests while iterating through the list, we do not want to exit the `for` cycle completely - if it happens we will drop the rest of the requests from the list and will never notice (I wouldn't like to debug such a silent error).

---

That's why we have try/except blocks for ;)